File: js/MaxSubArray.js
var maxSubArray = function(nums) {
  let max = nums[0], pre = max;
  for (let i = 1; i < nums.length; i++) {
    pre = Math.max(pre + nums[i], nums[i]);
    max = Math.max(max, pre);
    console.log(pre, max);
  }
  return max;
};

maxSubArray([-2,1,-3,4,-1,2,1,-5,4])

File: java/zijie-new/dp/MaxSubArray.java
package dp;

public class MaxSubArray {

    /**
     * 找到一个具有最大和的连续子数组 (一定是连续的)
     */
    public int maxSubArray(int[] nums) {
        if (nums.length == 0) return 0;
        // 表示前面的最大和是多少
        int[] dp = new int[nums.length + 1];
        int maxSum = Integer.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            if (dp[i] < 0) {
                dp[i + 1] = nums[i];
            } else {
                dp[i + 1] = dp[i] + nums[i];
            }
            maxSum = Math.max(maxSum, dp[i + 1]);
        }

        return maxSum;
    }
}
